### ISSUE TITLE
installation failure on R 3.5.0

### ISSUE THREAD

**Issue by steffengreup** [r-lib/vctrs#1153]
Unfortunately, installation of vctrs fails on my R 3.5.0 when installing the source version 0.3.1. Apparantly it has to do with a missing Altrep.h file:

``` r
> install.packages("vctrs")

Installing package into ‘H:/Apps/R_Library_nieuw’

(as ‘lib’ is unspecified)

  There is a binary version available but the source version is later:

      binary source needs_compilation

vctrs  0.2.4  0.3.1              TRUE

installing the source package ‘vctrs’

trying URL 'https://cran.rstudio.com/src/contrib/vctrs_0.3.1.tar.gz'

Content type 'application/x-gzip' length 946964 bytes (924 KB)

downloaded 924 KB

* installing *source* package 'vctrs' ...

** package 'vctrs' successfully unpacked and MD5 sums checked

** libs

*** arch - i386

C:/Rtools/mingw_32/bin/gcc  -I"C:/Program Files/R/R-35~1.0/include" -DNDEBUG          -O3 -Wall  -std=gnu99 -mtune=generic -c altrep-rle.c -o altrep-rle.o

In file included from altrep-rle.h:4:0,

                 from altrep-rle.c:2:

altrep.h:13:26: fatal error: R_ext/Altrep.h: No such file or directory

#include "R_ext/Altrep.h"

                          ^

compilation terminated.

make: *** [C:/Program Files/R/R-35~1.0/etc/i386/Makeconf:208: altrep-rle.o] Error 1

ERROR: compilation failed for package 'vctrs'
```

**Comment by steffengreup** [comment]
It seems that the Altrep.h file was missing in R 3.5.0: https://github.com/rwinlib/base/issues/9

After copying the Altrep.h file from a more recent R installation (3.5.2 in this case) to the R_ext folder in my R 3.5.0 installation, vctrs has no problem compiling.

This feels rather hacky though.

**Comment by lionel-** [comment]
Thanks for the diagnostic, it's helpful.

**Comment by hamgamb** [comment]
I am still unable to install vctrs 0.3.1 on R 3.5.0. Attempting to install the development version as I assume the commit hasn't made it to CRAN yet. 

``` r
devtools::install_github("r-lib/vctrs")
#> Using github PAT from envvar GITHUB_PAT
#> Downloading GitHub repo r-lib/vctrs@master
#> 
#>          checking for file 'C:\Users\gamb0043\AppData\Local\Temp\Rtmpm87HWx\remotes4298268a2ec9\r-lib-vctrs-f5f57e0/DESCRIPTION' ...  v  checking for file 'C:\Users\gamb0043\AppData\Local\Temp\Rtmpm87HWx\remotes4298268a2ec9\r-lib-vctrs-f5f57e0/DESCRIPTION' (439ms)
#>       -  preparing 'vctrs': (371ms)
#>    checking DESCRIPTION meta-information ...     checking DESCRIPTION meta-information ...   v  checking DESCRIPTION meta-information
#> -  cleaning src
#>       -  installing the package to process help pages
#>            -----------------------------------
#>   -  installing *source* package 'vctrs' ...
#>    ** libs
#>    C:/RBuildTools/3.5/mingw_64/bin/gcc  -I"C:/PROGRA~1/R/R-35~1.0/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mtune=generic -c altrep-rle.c -o altrep-rle.o
#>    In file included from altrep-rle.c:2:0:
#>    altrep-rle.h:23:8: error: unknown type name 'R_altrep_class_t'
#>     extern R_altrep_class_t altrep_rle_class;
#>            ^
#>    altrep-rle.c:20:1: error: unknown type name 'R_altrep_class_t'
#>     R_altrep_class_t altrep_rle_class;
#>     ^
#>    altrep-rle.c: In function 'altrep_rle_Make':
#>    altrep-rle.c:24:3: warning: implicit declaration of function 'R_new_altrep' [-Wimplicit-function-declaration]
#>       SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
#>       ^
#>    altrep-rle.c:24:14: warning: initialization makes pointer from integer without a cast
#>       SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
#>                  ^
#>    altrep-rle.c: In function 'vctrs_init_altrep_rle':
#>    altrep-rle.c:170:3: warning: implicit declaration of function 'R_make_altstring_class' [-Wimplicit-function-declaration]
#>       altrep_rle_class = R_make_altstring_class("altrep_rle", "vctrs", dll);
#>       ^
#>    altrep-rle.c:173:3: warning: implicit declaration of function 'R_set_altrep_Length_method' [-Wimplicit-function-declaration]
#>       R_set_altrep_Length_method(altrep_rle_class, altrep_rle_Length);
#>       ^
#>    altrep-rle.c:174:3: warning: implicit declaration of function 'R_set_altrep_Inspect_method' [-Wimplicit-function-declaration]
#>       R_set_altrep_Inspect_method(altrep_rle_class, altrep_rle_Inspect);
#>       ^
#>    altrep-rle.c:177:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_method' [-Wimplicit-function-declaration]
#>       R_set_altvec_Dataptr_method(altrep_rle_class, altrep_rle_Dataptr);
#>       ^
#>    altrep-rle.c:178:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_or_null_method' [-Wimplicit-function-declaration]
#>       R_set_altvec_Dataptr_or_null_method(altrep_rle_class, altrep_rle_Dataptr_or_null);
#>       ^
#> altrep-rle.c:179:3: warning: imp     altrep-rle.c:179:3: warning: implicit declaration of function 'R_set_altvec_Extract_subset_method' [-Wimplicit-function-declaration]
#>       R_set_altvec_Extract_subset_method(altrep_rle_class, altrep_rle_Extract_subset);
#>       ^
#>    altrep-rle.c:182:3: warning: implicit declaration of function 'R_set_altstring_Elt_method' [-Wimplicit-function-declaration]
#>       R_set_altstring_Elt_method(altrep_rle_class, altrep_rle_string_Elt);
#>       ^
#>    make: *** [C:/PROGRA~1/R/R-35~1.0/etc/x64/Makeconf:208: altrep-rle.o] Error 1
#>    ERROR: compilation failed for package 'vctrs'
#> -  removing 'C:/Users/gamb0043/AppData/Local/Temp/Rtmp4eL1Hc/Rinst356c7b692f/vctrs'
#>    In R CMD INSTALL
#>          -----------------------------------
#>      ERROR: package installation failed
#> 
#> Error: Failed to install 'vctrs' from GitHub:
#>   System command 'Rcmd.exe' failed, exit status: 1, stdout + stderr:
#> E> * checking for file 'C:\Users\gamb0043\AppData\Local\Temp\Rtmpm87HWx\remotes4298268a2ec9\r-lib-vctrs-f5f57e0/DESCRIPTION' ... OK
#> E> * preparing 'vctrs':
#> E> * checking DESCRIPTION meta-information ... OK
#> E> * cleaning src
#> E> * installing the package to process help pages
#> E>       -----------------------------------
#> E> * installing *source* package 'vctrs' ...
#> E> ** libs
#> E> C:/RBuildTools/3.5/mingw_64/bin/gcc  -I"C:/PROGRA~1/R/R-35~1.0/include" -DNDEBUG          -O2 -Wall  -std=gnu99 -mtune=generic -c altrep-rle.c -o altrep-rle.o
#> E> In file included from altrep-rle.c:2:0:
#> E> altrep-rle.h:23:8: error: unknown type name 'R_altrep_class_t'
#> E>  extern R_altrep_class_t altrep_rle_class;
#> E>         ^
#> E> altrep-rle.c:20:1: error: unknown type name 'R_altrep_class_t'
#> E>  R_altrep_class_t altrep_rle_class;
#> E>  ^
#> E> altrep-rle.c: In function 'altrep_rle_Make':
#> E> altrep-rle.c:24:3: warning: implicit declaration of function 'R_new_altrep' [-Wimplicit-function-declaration]
#> E>    SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
#> E>    ^
#> E> altrep-rle.c:24:14: warning: initialization makes pointer from integer without a cast
#> E>    SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
#> E>               ^
#> E> altrep-rle.c: In function 'vctrs_init_altrep_rle':
#> E> altrep-rle.c:170:3: warning: implicit declaration of function 'R_make_altstring_class' [-Wimplicit-function-declaration]
#> E>    altrep_rle_class = R_make_altstring_class("altrep_rle", "vctrs", dll);
#> E>    ^
#> E> altrep-rle.c:173:3: warning: implicit declaration of function 'R_set_altrep_Length_method' [-Wimplicit-function-declaration]
#> E>    R_set_altrep_Length_method(altrep_rle_class, altrep_rle_Length);
#> E>    ^
#> E> altrep-rle.c:174:3: warning: implicit declaration of function 'R_set_altrep_Inspect_method' [-Wimplicit-function-declaration]
#> E>    R_set_altrep_Inspect_method(altrep_rle_class, altrep_rle_Inspect);
#> E>    ^
#> E> altrep-rle.c:177:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_method' [-Wimplicit-function-declaration]
#> E>    R_set_altvec_Dataptr_method(altrep_rle_class, altrep_rle_Dataptr);
#> E>    ^
#> E> altrep-rle.c:178:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_or_null_method' [-Wimplicit-function-declaration]
#> E>    R_set_altvec_Dataptr_or_null_method(altrep_rle_class, altrep_rle_Dataptr_or_null);
#> E>    ^
#> E> altrep-rle.c:179:3: warning: implicit declaration of function 'R_set_altvec_Extract_subset_method' [-Wimplicit-function-declaration]
#> E>    R_set_altvec_Extract_subset_method(altrep_rle_class, altrep_rle_Extract_subset);
#> E>    ^
#> E> altrep-rle.c:182:3: warning: implicit declaration of function 'R_set_altstring_Elt_method' [-Wimplicit-function-declaration]
#> E>    R_set_altstring_Elt_method(altrep_rle_class, altrep_rle_string_Elt);
#> E>    ^
#> E> make: *** [C:/PROGRA~1/R/R-35~1.0/etc/x64/Makeconf:208: altrep-rle.o] Error 1
#> E> ERROR: compilation failed for package 'vctrs'
#> E> * removing 'C:/Users/gamb0043/AppData/Local/Temp/Rtmp4eL1Hc/Rinst356c7b692f/vctrs'
#> E> In R CMD INSTALL
#> E>       -----------------------------------
#> E> ERROR: package installation failed
```

<sup>Created on 2020-07-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by lionel-** [comment]
I conditioned in the .h file but not the .c file 😬

**Comment by lionel-** [comment]
@hamgamb Can you try again please?

**Comment by hamgamb** [comment]
@lionel- working now. Thanks 👍

**Comment by OFlanagan** [comment]
I am still encountering this issue:
install.packages("vctrs")
Installing package into ‘C:/Users/Owen/Documents/R/win-library/3.5’
(as ‘lib’ is unspecified)

  There is a binary version available but the source version is later:
      binary source needs_compilation
vctrs  0.2.4  0.3.2              TRUE

installing the source package ‘vctrs’

trying URL 'https://cran.rstudio.com/src/contrib/vctrs_0.3.2.tar.gz'
Content type 'application/x-gzip' length 958372 bytes (935 KB)
downloaded 935 KB

* installing *source* package 'vctrs' ...
** package 'vctrs' successfully unpacked and MD5 sums checked
** libs

*** arch - i386
C:/RBuildTools/3.4/mingw_32/bin/gcc  -I"C:/PROGRA~1/R/R-35~1.0/include" -DNDEBUG          -O3 -Wall  -std=gnu99 -mtune=generic -c altrep-rle.c -o altrep-rle.o
In file included from altrep-rle.c:2:0:
altrep-rle.h:23:8: error: unknown type name 'R_altrep_class_t'
 extern R_altrep_class_t altrep_rle_class;
        ^
altrep-rle.c:20:1: error: unknown type name 'R_altrep_class_t'
 R_altrep_class_t altrep_rle_class;
 ^
altrep-rle.c: In function 'altrep_rle_Make':
altrep-rle.c:24:3: warning: implicit declaration of function 'R_new_altrep' [-Wimplicit-function-declaration]
   SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
   ^
altrep-rle.c:24:14: warning: initialization makes pointer from integer without a cast
   SEXP res = R_new_altrep(altrep_rle_class, input, R_NilValue);
              ^
altrep-rle.c: In function 'vctrs_init_altrep_rle':
altrep-rle.c:170:3: warning: implicit declaration of function 'R_make_altstring_class' [-Wimplicit-function-declaration]
   altrep_rle_class = R_make_altstring_class("altrep_rle", "vctrs", dll);
   ^
altrep-rle.c:173:3: warning: implicit declaration of function 'R_set_altrep_Length_method' [-Wimplicit-function-declaration]
   R_set_altrep_Length_method(altrep_rle_class, altrep_rle_Length);
   ^
altrep-rle.c:174:3: warning: implicit declaration of function 'R_set_altrep_Inspect_method' [-Wimplicit-function-declaration]
   R_set_altrep_Inspect_method(altrep_rle_class, altrep_rle_Inspect);
   ^
altrep-rle.c:177:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_method' [-Wimplicit-function-declaration]
   R_set_altvec_Dataptr_method(altrep_rle_class, altrep_rle_Dataptr);
   ^
altrep-rle.c:178:3: warning: implicit declaration of function 'R_set_altvec_Dataptr_or_null_method' [-Wimplicit-function-declaration]
   R_set_altvec_Dataptr_or_null_method(altrep_rle_class, altrep_rle_Dataptr_or_null);
   ^
altrep-rle.c:179:3: warning: implicit declaration of function 'R_set_altvec_Extract_subset_method' [-Wimplicit-function-declaration]
   R_set_altvec_Extract_subset_method(altrep_rle_class, altrep_rle_Extract_subset);
   ^
altrep-rle.c:182:3: warning: implicit declaration of function 'R_set_altstring_Elt_method' [-Wimplicit-function-declaration]
   R_set_altstring_Elt_method(altrep_rle_class, altrep_rle_string_Elt);
   ^
make: *** [altrep-rle.o] Error 1
ERROR: compilation failed for package 'vctrs'
* removing 'C:/Users/Owen/Documents/R/win-library/3.5/vctrs'
In R CMD INSTALL
Warning in install.packages :
  installation of package ‘vctrs’ had non-zero exit status

**Comment by hamgamb** [comment]
@OFlanagan - my understanding is that the fix for this issue (https://github.com/r-lib/vctrs/commit/154432bfde9ada65c352824df695b13ce08ac913) was made after the 0.3.2 version was released to CRAN. If you can install the development version instead with `devtools::install_github("r-lib/vctrs")` it should work for you.

**Comment by lionel-** [comment]
That's right. But we made a mistake in a compat layer in dplyr for the dev version, so if you install it you'll also need that dplyr fix: https://github.com/tidyverse/dplyr/pull/5464

```r
devtools::install_github("r-lib/vctrs")
devtools::install_github("lionel-/dplyr@fix-proxy-order)
```

Really the best thing is to install the patched version of the 3.5 series of R. The latest one is 3.5.3 and it includes the fix to the ALTREP headers on Windows, which caused the compilation failures. https://cran.r-project.org/bin/windows/base/old/3.5.3/

Or, of course, install the very latest version of the 4.0 series.
